compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala

```diff
@@ -314,56 +314,46 @@ object messages {
-        .filter(_.isType == name.isTypeName)
-        .flatMap { sym =>
-          if (sym.flagsUNSAFE.isOneOf(Synthetic | PrivateLocal) || sym.isConstructor) Nil
-          else List((sym.name.show, sym))
-        }
+      val maxDist = 3  // maximal number of differences to be considered for a hint
+      val missing = name.show
+
+      // The names of all non-synthetic, non-private members of `site`
+      // that are of the same type/term kind as the missing member.
+      def candidates: Set[String] =
+        for
+          bc <- site.widen.baseClasses.toSet
+          sym <- bc.info.decls.filter(sym =>
+            sym.isType == name.isTypeName
+            && !sym.isConstructor
+            && !sym.flagsUNSAFE.isOneOf(Synthetic | Private))
+        yield sym.name.show
 
       // Calculate Levenshtein distance
-      def distance(n1: Iterable[?], n2: Iterable[?]) =
-        n1.foldLeft(List.range(0, n2.size)) { (prev, x) =>
-          (prev zip prev.tail zip n2).scanLeft(prev.head + 1) {
-            case (h, ((d, v), y)) => math.min(
-              math.min(h + 1, v + 1),
-              if (x == y) d else d + 1
-            )
-          }
-        }.last
-
-      // Count number of wrong characters
-      def incorrectChars(x: (String, Int, Symbol)): (String, Symbol, Int) = {
-        val (currName, _, sym) = x
-        val matching = name.show.zip(currName).foldLeft(0) {
-          case (acc, (x,y)) => if (x != y) acc + 1 else acc
-        }
-        (currName, sym, matching)
-      }
-
-      // Get closest match in `site`
-      def closest: List[String] =
-        decls
-        .map { (n, sym) => (n, distance(n, name.show), sym) }
-        .collect {
-          case (n, dist, sym)
-          if dist <= maxDist && dist < (name.toString.length min n.length) =>
-            (n, dist, sym)
-        }
-        .groupBy(_._2).toList
-        .sortBy(_._1)
-        .headOption.map(_._2).getOrElse(Nil)
-        .map(incorrectChars).toList
-        .sortBy(_._3)
-        .map(_._1)
-          // [Martin] Note: I have no idea what this does. This shows the
-          // pitfalls of not naming things, functional or not.
+      def distance(s1: String, s2: String): Int =
+        val dist = Array.ofDim[Int](s2.length + 1, s1.length + 1)
+        for
+          j <- 0 to s2.length
+          i <- 0 to s1.length
+        do
+          dist(j)(i) =
+            if j == 0 then i
+            else if i == 0 then j
+            else if s2(j - 1) == s1(i - 1) then dist(j - 1)(i - 1)
+            else (dist(j - 1)(i) min dist(j)(i - 1) min dist(j - 1)(i - 1)) + 1
+        dist(s2.length)(s1.length)
+
+      // A list of possible candidate strings with their Levenstein distances
+      // to the name of the missing member
+      def closest: List[(Int, String)] = candidates
+        .toList
+        .map(n => (distance(n.show, missing), n))
+        .filter((d, n) => d <= maxDist && d < missing.length & d < n.length)
```

Comment:
Is that `&` instead of `&&` intentional ?

Reply:
No, that's a typo

Reply:
"Did you mean &&? or did you subconsciously intend to reverse the semantics of & and && in Scala 3 because that would look nicer?"